### PR TITLE
1605 pass wormconfig options

### LINKED ISSUE #1605: Implement exploitation in Puppet
Implement the `exploit_host()` function of the concrete Puppet. Modify the Master to pass the appropriate options to the exploiters (i.e. propagation credentials). The `smb` exploit may require additional options.

**Note**
There's currently a bug where MSSQL exploiter can fail to propagate and prevent other exploiters from running. We can most likely resolve this by having each exploiter return two booleans, one to indicate whether or not the victim is vulnerable and another to indicate whether or not propagation was successful. This would also let us simplify the exploiters config, which now have a true/false field to indicate whether or not an exploiter is capable of propagating.

## Tasks

- [x] Modify config to remove boolean "propagator" field from config (0d) @ilija-lazoroski 
- [x] Add a periodic check (between hosts and/or between exploiters) to get updated credentials from the Island (0d) @mssalvatore 
- [x] Pass other WormConfig settings as options (0d) @ilija-lazoroski 
    - [x] Target dropper paths
    - [x] Pass credentials as options to exploiters (See #1721 for a starting point)
- [x] Copy commonly used functions on WormConfig to a utils/tools module and add tests (0d) @mssalvatore
    - [x] get_exploit_user_password_or_hash_product()
    - [x] get_exploit_user_ssh_key_pairs()
    - [x] get_exploit_user_password_pairs()
- [x] Modify ExploitResultData (0d) @shreyamalviya
    - [x] Add OS
    - [x] Add `exploit_successful`
    - [x] Add `propagation_successful`
    - [x] Modify mock_puppet
- [x] Remove `skip_if_exploit_file_exists` option from config (0d) @shreyamalviya
- [x] SSH (0d) @ilija-lazoroski 
    - [x] Don't modify VictimHost object
    - [x] Change attempt logging and remove credential hashes from log statements
    - [x] Return ExploitResultData
    - [x] Call new `get_exploit_user*()` functions instead of `self._config.*()`
    - [x] Accept ITelemetryMessenger in `__init__()` and use it to send telemetries
- [x] Hadoop (0d) @shreyamalviya 
    - [x] Remove code that sets architecture from web_rce.py
    - [x] Return ExploitResultData
    - [x] Accept ITelemetryMessenger in `__init__()` and use it to send telemetries
    - [x] Wrap ~Struts~Hadoop with wrapper and load into the puppet in `monkey.py`
- [x] Remove dependency on `Plugin` from `HostExploiter` (0d) @mssalvatore 
- [x] Implement `exploit_host()` in `Puppet` (0d) @mssalvatore 
    - [x] Add a wrapper around Exploiters that accepts the necessary parameters for exploitation and object construction (VictimHost, Telemetry messenger, options)

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do? 

Pass wormconfig options, related to #1605 . 

Add any further explanations here. 

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] ~Is the TravisCI build passing?~
* [ ] ~Was the CHANGELOG.md updated to reflect the changes?~
* [ ] ~Was the documentation framework updated to reflect the changes?~

## Testing Checklist

* [ ] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by {Running the Monkey locally with relevant config/running Island/...} 
* [x] If applicable, add screenshots or log transcripts of the feature working

## Explain Changes

Are the commit messages enough? If not, elaborate.

### CODE REVIEW

**Review by mssalvatore — CHANGES_REQUESTED** [review]

**Review by VakarisZ — APPROVED** [review]

File: monkey/infection_monkey/master/exploiter.py [hunk 1]
```diff
@@ -49,6 +49,27 @@ def exploit_hosts(
...
         )
 
+    @staticmethod
+    def _process_exploiter_config(exploiter_config: Mapping) -> List[Mapping]:
+        # Run vulnerability exploiters before brute force exploiters to minimize the effect of
+        # account lockout due to invalid credentials
+        ordered_exploiters = chain(
+            exploiter_config["vulnerability"], exploiter_config["brute_force"]
+        )
+        exploiters_to_run = list(deepcopy(ordered_exploiters))
```
Comment (shreyamalviya) [inline monkey/infection_monkey/master/exploiter.py:59]: Why do we need to convert an iterable to a list? As far as I can tell, we should be able to use it as it is.

File: monkey/infection_monkey/master/exploiter.py [hunk 2]
```diff
@@ -49,6 +49,27 @@ def exploit_hosts(
...
+
+        for exploiter in exploiters_to_run:
+            # This order allows exploiter-specific options to
+            # override general options for all exploiters.
+            exploiter["options"] = {**exploiter_config["options"], **exploiter["options"]}
+
+        logger.debug(
+            "Agent is configured to run the following exploiters in order: "
+            f"{','.join([e['name'] for e in exploiters_to_run])}"
```
Comment (shreyamalviya) [inline monkey/infection_monkey/master/exploiter.py:68]:
```suggestion
            f"{', '.join([e['name'] for e in exploiters_to_run])}"
```

File: monkey/infection_monkey/master/exploiter.py (hunk 1 above)
Comment (mssalvatore) [reply monkey/infection_monkey/master/exploiter.py:59]: Iterators can only be used once, so after the loop on lines 61-64, the iterator would be exhausted and all we could return would be an empty iterator.

Comment (shreyamalviya) [reply monkey/infection_monkey/master/exploiter.py:59]: Oh right. Thanks.

**Review by shreyamalviya — APPROVED** [review]

**Pull request merged**